core/io/logger.cpp

```diff
@@ -266,3 +267,357 @@ CompositeLogger::~CompositeLogger() {
+	// any buffered callables that get attached might start getting messages midway through
+	// we're ok with that, it's still a chronologically coherent block
+	for (const Dictionary &log : buffered_logs_mirror) {
+		// This is the same index-as-iterator-to-avoid-lock-issues dance we do in register_log_capture_non_thread_safe()
+		int index_to_send_to = 0;
+
+		while (true) {
+			Callable callable;
+			{
+				MutexLock lock(mutex);
+				if (index_to_send_to >= captures_buffered.size()) {
+					// we done!
+					break;
+				}
+
+				callable = captures_buffered[index_to_send_to++];
+			}
+
+			dispatch_message(log, callable);
+		}
+	}
+
+	if (pre_buffering.is_set()) {
+		// Buffering is done!
+		// Read the comment near the bottom of recalculate_state for a possible race condition and how it's dealt with.
+		// IT IS REALLY IMPORTANT THAT THIS HAPPENS ONLY ON THE MAIN THREAD.
+		MutexLock lock(mutex);
+		pre_buffering.clear();
+		recalculate_state();
+	}
+}
+
+void UserLogManagerLogger::process(const Dictionary &p_message) {
+	Vector<Callable> captures_non_thread_safe_mirror;
+	{
+		// Shove another item into the buffer, if we need to.
+
+		// We also atomically grab `captures_non_thread_safe` to avoid race conditions, in the extraordinarily unlikely case where:
+		// * this function buffers a message, releases the lock, and is immediately preempted
+		// * in another thread, an already-running register_log_capture_non_thread_safe dispatches our new message from the buffer
+		// * that same register_log_capture_non_thread_safe gets to the end of the buffer and adds itself to captures_non_thread_safe
+		// * this function resumes, reaches the captures_non_thread_safe section, and sends the same message again
+
+		// register_log_capture_non_thread_safe intentionally finishes the buffered log message replay atomically with adding a new handler to the vector
+		// so we do the same thing here, adding a message to the log atomically with copying the handlers
+		// Vector COW behavior prevents this from being expensive in the common case
+		MutexLock lock(mutex);
+		if (state.get() == STATE_BUFFERING) {
+			buffered_logs.append(p_message);
+		}
+		captures_non_thread_safe_mirror = captures_non_thread_safe;
+	}
+
+	// Dispatch to all the non_thread_safe callables at the moment we added to the buffer
+	// We actually don't have to care about cutesy thread-safety for once because we're working off a local copy of the captures list
+
+	for (const Callable &callable : captures_non_thread_safe_mirror) {
+		dispatch_message(p_message, callable);
+	}
+}
+
+void UserLogManagerLogger::dispatch_message(const Dictionary &p_message, const Callable &p_callable) {
+	// ideally we should verify that the mutex is *not* held by this thread, but the current API provides no way to do that
+	// (it's possible it's already been scooped up by another thread, that's fine)
+
+	if (p_callable.is_null()) {
```

Comment:
```suggestion
	if (!p_callable.is_valid()) {
```

Reply:
No, it doesn't check if it's been deleted, unless you mean specifically set to `Callable()`, if the callable has an object that was freed for example 

It's not significantly slower, and ensures things won't crash, it's not correct to check this unless you're certain the object isn't freed

Reply:
Yes, I mean specifically set to `Callable()`. That's what `unregister_callable()` does.

I'm surprised `callp()` doesn't protect against this, but, sure, I'll change that.

Reply:
It does in debug builds, but not otherwise, so it's important to use the correct check